### PR TITLE
Add a new "fast" shader path that doesn't use texelFetch in glsl

### PR DESCRIPTION
**Pull request by fzurita**
As the title says and as discussed here: https://github.com/gonetz/GLideN64/issues/2542

### CODE REVIEW

**Comment by gonetz** [comment]
Hi,

I tested that branch (VS 2017). It works fine. I can switch between fast and accurate modes on fly. That's great.
What is not great: lots of code is duplicated with the proposed approach.
I strongly against code duplication, even if we want completely separate two code paths.

I suggest to use another architectural solution, the one, which is already used for DisplayWindow.
Base class implements some high-level logic. Low-level implementation is provided via private abstract virtual methods.
There are specialized implementations of DisplayWindow: DisplayWindowWindows and DisplayWindowMupen64plus.

For CombinerProgramUniformFactory it can be done like this:

```
class CombinerProgramUniformFactory
{
public:
	CombinerProgramUniformFactory(const opengl::GLInfo & _glInfo);
	virtual ~CombinerProgramUniformFactory();

	void buildUniforms(GLuint _program,
						const CombinerInputs & _inputs,
						const CombinerKey & _key,
						UniformGroups & _uniforms);

protected:
	const opengl::GLInfo & m_glInfo;

private:
  virtual void _addNoiseTex(GLuint _program, UniformGroups & _uniforms) const = 0;
  virtual void _addScreenSpaceTriangleInfo(GLuint _program, UniformGroups & _uniforms) const = 0;
...
  virtual void _addTextureSize(GLuint _program, UniformGroups & _uniforms) const = 0;
  virtual void _addClampWrapMirrorEngine(GLuint _program, UniformGroups & _uniforms) const = 0;
...
  virtual void _addLights(GLuint _program, UniformGroups & _uniforms) const = 0;
};

void CombinerProgramUniformFactory::buildUniforms(GLuint _program,
const CombinerInputs & _inputs,
const CombinerKey & _key,
UniformGroups & _uniforms)
{
	_addNoiseTex(_program, _uniforms);
	_addScreenSpaceTriangleInfo(_program, _uniforms);
...
	if (_inputs.usesHwLighting())
		_addLights(_program, _uniforms);
}

class CombinerProgramUniformFactoryCommon: public CombinerProgramUniformFactory
{
private:
  void _addNoiseTex(GLuint _program, UniformGroups & _uniforms) const override;
  void _addScreenSpaceTriangleInfo(GLuint _program, UniformGroups & _uniforms) const override;
...
  void _addLights(GLuint _program, UniformGroups & _uniforms) const override;
};

void CombinerProgramUniformFactoryCommon::_addNoiseTex(GLuint _program, UniformGroups & _uniforms)
{
	_uniforms.emplace_back(new UNoiseTex(_program));
}

class CombinerProgramUniformFactoryFast: public CombinerProgramUniformFactoryCommon
{
public:
	CombinerProgramUniformFactoryFast(const opengl::GLInfo & _glInfo);
	
private:
  void _addTextureSize(GLuint _program, UniformGroups & _uniforms) const override;
  void _addClampWrapMirrorEngine(GLuint _program, UniformGroups & _uniforms) const override;
...	
};

class CombinerProgramUniformFactoryAccurate: public CombinerProgramUniformFactoryCommon
{
public:
	CombinerProgramUniformFactoryAccurate(const opengl::GLInfo & _glInfo);
	
private:
  void _addTextureSize(GLuint _program, UniformGroups & _uniforms) const override;
  void _addClampWrapMirrorEngine(GLuint _program, UniformGroups & _uniforms) const override;
...	
};

```
Since accurate mode does not use UMipmap2, 
```
void CombinerProgramUniformFactoryAccurate::_addMipmap2(GLuint _program, UniformGroups & _uniforms) const
{}
```

Uniform structures, such as iUniform, fUniform and so on, should be moved to CombinerProgramUniformFactoryCommon.h/.cpp

For CombinerProgramBuilder is can be done similarly:

```
class CombinerProgramBuilder
{
public:
	CombinerProgramBuilder(const opengl::GLInfo & _glinfo, opengl::CachedUseProgram * _useProgram);
	virtual ~CombinerProgramBuilder();

	graphics::CombinerProgram * buildCombinerProgram(Combiner & _color, Combiner & _alpha, const CombinerKey & _key);

	virtual const ShaderPart * getVertexShaderHeader() const = 0;

	virtual const ShaderPart * getFragmentShaderHeader() const = 0;

	virtual const ShaderPart * getFragmentShaderEnd() const = 0;

	virtual bool isObsolete() const = 0;

private:
	CombinerInputs compileCombiner(const CombinerKey & _key, Combiner & _color, Combiner & _alpha, std::string & _strShader);
	virtual void _writeBlender1(std::stringstream& ssShader) const = 0;
	virtual void _writeBlender2(std::stringstream& ssShader) const = 0;
	virtual void _writeBlenderAlpha(std::stringstream& ssShader) const = 0;
...
	virtual void _writeCoverage(std::stringstream& ssShader) const = 0;

	std::unique_ptr<CombinerProgramUniformFactory> m_uniformFactory;
	GLuint  m_vertexShaderRect;
	GLuint  m_vertexShaderTriangle;
	GLuint  m_vertexShaderTexturedRect;
	GLuint  m_vertexShaderTexturedTriangle;
	opengl::CachedUseProgram * m_useProgram;
	u32 m_combinerOptionsBits;
	bool m_useCoverage = false;
};

CombinerInputs CombinerProgramBuilder::compileCombiner(const CombinerKey & _key, Combiner & _color, Combiner & _alpha, std::string & _strShader)
{
...
	if (combinedAlphaC(combine))
		_writeSignExtendAlphaC(ssShader);
	else if (combinedAlphaABD(combine))
		_writeSignExtendAlphaABD(ssShader);
...
}

class CombinerProgramBuilderCommon: public CombinerProgramBuilder
{
public
	using ShaderPartPtr = std::unique_ptr<ShaderPart>;

private:
	void _writeBlender1(std::stringstream& ssShader) const;
	void _writeBlender2(std::stringstream& ssShader) const;
	void _writeBlenderAlpha(std::stringstream& ssShader) const;
...
	void _writeCoverage(std::stringstream& ssShader) const;
	
	ShaderPartPtr m_blender1;
	ShaderPartPtr m_blender2;
	...
};

void CombinerProgramBuilderCommon::_writeBlender1(std::stringstream& ssShader) const
{
	m_blender1->write(ssShader);
}

class CombinerProgramBuilderFast: public CombinerProgramBuilderCommon
{
public:
	CombinerProgramBuilderFast(const opengl::GLInfo & _glinfo, opengl::CachedUseProgram * _useProgram);

private:
	void _writeShaderFragmentGlobalVariablesTex(std::stringstream& ssShader) const;
	void _writeClampWrapMirrorEngine(std::stringstream& ssShader) const;
...

	ShaderPartPtr m_fragmentGlobalVariablesTex;
...	
};

class CombinerProgramBuilderAccurate: public CombinerProgramBuilderCommon
{
public:
	CombinerProgramBuilderAccurate(const opengl::GLInfo & _glinfo, opengl::CachedUseProgram * _useProgram);

private:
	void _writeShaderFragmentGlobalVariablesTex(std::stringstream& ssShader) const;
	void _writeClampWrapMirrorEngine(std::stringstream& ssShader) const;
...

	ShaderPartPtr m_fragmentGlobalVariablesTex;
...	
};

```
This way the code duplication will be minimized.

**Comment by fzurita** [comment]
Yeah, I understand the code duplication concern and it would be good to minimize it. Your proposed path forward seems reasonable. It will take a good amount of work to implement though. I will look into doing it.

**Comment by fzurita** [comment]
Well, I did it, but I messed up somewhere:

https://github.com/fzurita/GLideN64/tree/optional_fast_shader_path_refactor

Everything runs, but 3D is just black. There are no GL errors to give me a hint where the problem is. I'm going to try to back track to see where the problem is.

**Comment by fzurita** [comment]
Ok, I figured it out, I had ```CombinerProgramUniformFactoryCommon::_addViewportInfo``` using the wrong uniformGroup. I updated the pull request, let me know what you think. This took a good amount of work.

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.h [hunk 1]
```diff
@@ -2,108 +2,122 @@
...
-		const ShaderPart * getFragmentShaderHeader() const;
+		virtual const ShaderPart * getFragmentShaderHeader() const = 0;
 
-		const ShaderPart * getFragmentShaderEnd() const;
+		virtual const ShaderPart * getFragmentShaderEnd() const = 0;
 
-		bool isObsolete() const;
+		virtual bool isObsolete() const = 0;
+
+		static u32 cycleType;
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.h:57]: s_cycleType

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.h [hunk 2]
```diff
@@ -2,108 +2,122 @@
...
+		static TextureConvert textureConvert;
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.h:58]: s_textureConvert

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.h [hunk 3]
```diff
@@ -2,108 +2,122 @@
...
+		virtual const ShaderPart * getVertexShaderHeader() const = 0;
 
-		const ShaderPart * getFragmentShaderHeader() const;
+		virtual const ShaderPart * getFragmentShaderHeader() const = 0;
 
-		const ShaderPart * getFragmentShaderEnd() const;
+		virtual const ShaderPart * getFragmentShaderEnd() const = 0;
 
-		bool isObsolete() const;
+		virtual bool isObsolete() const = 0;
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.h:55]: Indent is unnecessary here. Please remove left indent for the code inside `namespace `brackets, as it is done in cpp file.

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderAccurate.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
+#pragma once
+#include <memory>
+#include <Combiner.h>
+#include <Graphics/OpenGLContext/opengl_GLInfo.h>
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderAccurate.h:4]: redundant includes

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderAccurate.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		ShaderPartPtr m_fragmentReadTexCopyMode;
+		ShaderPartPtr m_fragmentReadTex0;
+		ShaderPartPtr m_fragmentReadTex1;
+		ShaderPartPtr m_shaderMipmap;
+		ShaderPartPtr m_shaderReadtex;
+		ShaderPartPtr m_shaderReadtexCopyMode;
+		ShaderPartPtr m_shaderTextureEngine;
+		ShaderPartPtr m_fragmentTextureEngineTex0;
+		ShaderPartPtr m_fragmentTextureEngineTex1;
+	};
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderAccurate.h:60]: Indent is unnecessary here. Please remove left indent for the code inside `namespace `brackets, as it is done in cpp file.

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h [hunk 6]
```diff
@@ -0,0 +1,133 @@
+#pragma once
+#include <memory>
+#include <Combiner.h>
+#include <Graphics/OpenGLContext/opengl_GLInfo.h>
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h:4]: redundant includes

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderAccurate.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	class CombinerProgram;
+}
+
+namespace opengl {
+	class CachedUseProgram;
+}
+
+namespace glsl {
+	class CombinerInputs;
+}
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderAccurate.h:17]: at least part of it was already declared in included headers

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	class CombinerProgram;
+}
+
+namespace opengl {
+	class CachedUseProgram;
+}
+
+namespace glsl {
+	class CombinerInputs;
+}
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h:17]: at least part of it was already declared

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+		ShaderPartPtr m_shaderN64DepthRender;
+		ShaderPartPtr m_shaderCoverage;
+
+		u32 m_combinerOptionsBits;
+
+		GLuint m_vertexShaderRect;
+		GLuint m_vertexShaderTriangle;
+		GLuint m_vertexShaderTexturedRect;
+		GLuint m_vertexShaderTexturedTriangle;
+	};
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h:131]: Indent is unnecessary here. Please remove left indent for the code inside `namespace `brackets, as it is done in cpp file.

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderFast.h [hunk 10]
```diff
@@ -0,0 +1,61 @@
+#pragma once
+#include <memory>
+#include <Combiner.h>
+#include <Graphics/OpenGLContext/opengl_GLInfo.h>
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderFast.h:4]: redundant includes

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderFast.h [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	class CombinerProgram;
+}
+
+namespace opengl {
+	class CachedUseProgram;
+}
+
+namespace glsl {
+	class CombinerInputs;
+}
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderFast.h:17]: at least part of it was already declared

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderFast.h [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+		ShaderPartPtr m_fragmentReadTex0;
+		ShaderPartPtr m_fragmentReadTex1;
+		ShaderPartPtr m_fragmentClampWrapMirrorTex0;
+		ShaderPartPtr m_fragmentClampWrapMirrorTex1;
+		ShaderPartPtr m_fragmentReadTexCopyMode;
+		ShaderPartPtr m_shaderMipmap;
+		ShaderPartPtr m_shaderReadtex;
+		ShaderPartPtr m_shaderReadtexCopyMode;
+		ShaderPartPtr m_shaderClampWrapMirror;
+	};
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderFast.h:59]: Indent is unnecessary here. Please remove left indent for the code inside `namespace` brackets, as it is done in cpp file.

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+		void _writeFragmentRenderTarget(std::stringstream& ssShader) const override;
+		void _writeShaderFragmentMainEnd(std::stringstream& ssShader) const override;
+		void _writeShaderCalcLight(std::stringstream& ssShader) const override;
+		void _writeShaderNoise(std::stringstream& ssShader) const override;
+		void _writeShaderDither(std::stringstream& ssShader) const override;
+		void _writeShaderWriteDepth(std::stringstream& ssShader) const override;
+		void _writeShaderN64DepthCompare(std::stringstream& ssShader) const override;
+		void _writeShaderN64DepthRender(std::stringstream& ssShader) const override;
+
+		typedef std::unique_ptr<ShaderPart> ShaderPartPtr;
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilderCommon.h:81]: better
`using ShaderPartPtr = std::unique_ptr<ShaderPart>;`
and place it in public section of glsl::CombinerProgramBuilder class.
the same for CombinerProgramBuilderFast and CombinerProgramBuilderAccurate

File: src/Graphics/OpenGLContext/GLSL/glsl_ShaderStorage.cpp [hunk 14]
```diff
@@ -336,7 +338,13 @@ bool ShaderStorage::loadShadersStorage(graphics::Combiners & _combiners)
...
 
 		displayLoadProgress(L"LOAD COMBINER SHADERS %.1f%%", 0.0f);
-		CombinerProgramUniformFactory uniformFactory(m_glinfo);
+		std::unique_ptr<CombinerProgramUniformFactory> uniformFactory;
+
+		if (config.generalEmulation.enableFastInaccurateShaders) {
+			uniformFactory = std::make_unique<CombinerProgramUniformFactoryFast>(m_glinfo);
+		} else {
+			uniformFactory = std::make_unique<CombinerProgramUniformFactoryAccurate>(m_glinfo);
+		}
```
[the remark this answers is not in the document]
Comment (fzurita) [reply src/Graphics/OpenGLContext/GLSL/glsl_ShaderStorage.cpp]: Unfortunately ternary operator doesn't seem to work well with this: https://stackoverflow.com/questions/29842095/incompatible-operand-types-when-using-ternary-conditional-operator

**Comment by gonetz** [comment]
Does it work for you? I got incomplete vertex shader in SpecialShadersFactory::createTextDrawerShader()

**Comment by fzurita** [comment]
Hmm, interesting, I did not re-test after making the requested changes since they seemed fairly minor. It's possible I broke something along the way.

**Comment by gonetz** [comment]
https://github.com/gonetz/GLideN64/tree/fzurita-optional_fast_shader_path

**Comment by fzurita** [comment]
The Linux build seems to be failing.

File: src/Textures.cpp [hunk 15]
```diff
@@ -1180,15 +1363,15 @@ void TextureCache::_load(u32 _tile, CachedTexture *_pTexture)
...
 
 			getLoadParams(tmptex.format, tmptex.size);
-			_getTextureDestData(tmptex, texData.get() + texDataOffset, glInternalFormat, GetTexel, &line);
+			_getTextureDestData(tmptex, &m_tempTextureHolder[texDataOffset], glInternalFormat, GetTexel, &line);
 
 			if (m_toggleDumpTex &&
 				config.textureFilter.txHiresEnable != 0 &&
 				config.hotkeys.enabledKeys[Config::HotKey::hkTexDump] != 0) {
-				txfilter_dmptx((u8*)texData.get() + texDataOffset, tmptex.width, tmptex.height,
+				txfilter_dmptx((u8*)m_tempTextureHolder.data() + texDataOffset, tmptex.width, tmptex.height,
```
[the remark this answers is not in the document]
Comment (fzurita) [reply src/Textures.cpp:1374]: Ok, I can fix it now since I'm touching that code anyways.

[the remark this answers is not in the document]
Comment (fzurita) [reply src/Textures.cpp]: Ok, fixed.

[the remark this answers is not in the document]
Comment (gonetz) [reply src/Textures.cpp]: I already force-pushed the branch

[the remark this answers is not in the document]
Comment (gonetz) [reply src/Textures.cpp]: But I did only  two corrections: this one and a correction in glsl_CombinerProgramBuilderFast.cpp
```
-namespace glsl {
+namespace {
+using namespace glsl;
```
So you can do these fixes in your branch.

[the remark this answers is not in the document]
Comment (fzurita) [reply src/Textures.cpp]: It seems like I already have the change in glsl_CombinerProgramBuilderFast.cpp?

**Comment by gonetz** [comment]
> I finished renaming the config parameter to "Inaccurate Texture Coordinates"

Thanks!

**Comment by fzurita** [comment]
@gonetz Something went wrong, this is still wrong :

https://github.com/gonetz/GLideN64/blob/4c7b17db04d09dc0443219270c2b20e510c5c83e/src/Textures.cpp#L1374-L1377

**Comment by fzurita** [comment]
@gonetz This is a real issue... I missed two instances when I renamed the config parameter:

https://github.com/gonetz/GLideN64/blob/4c7b17db04d09dc0443219270c2b20e510c5c83e/src/GLideNUI/Settings.cpp#L51

https://github.com/gonetz/GLideN64/blob/4c7b17db04d09dc0443219270c2b20e510c5c83e/src/GLideNUI/Settings.cpp#L197

**Pull request closed without merging**